src/framework/constraints.ml

```diff
@@ -21,6 +21,7 @@ struct
 
   let name () = S.name () ^" hashconsed"
 
+  type marshal = S.marshal (* TODO: should hashcons table be in here to avoid relift altogether? *)
```

Comment:
Not sure if this would help, this table would be completely disconnected from the marshaled solution, right?

Reply:
If spec marshaled data goes together with solution, then it's no longer disconnected. I didn't try this but just thought that maybe if all hashconsing tables are marshaled (not sure if they can though, they use weak data structures internally) and restored, all the solver data is already consistent with hashconsing and nothing needs to be relifted.

Reply:
>if all hashconsing tables are marshaled (not sure if they can though, they use weak data structures internally)

It does indeed not look like this is possible, as the implementation of `BatHashcons.Table` uses `Weak`: 
https://github.com/ocaml-batteries-team/batteries-included/blob/99dcf7d66a6476dcd04f800acfb43b3af09a1d0e/src/batHashcons.ml#L64-L68
and those cannot be marshaled (see here: https://ocaml.org/api/Weak.html, the documentation is referring to the `Marshal` module by the standard library, but you also get an error `Fatal error: exception Invalid_argument("output_value: abstract value (Abstract)")` if your try to marshal the hashcons-tables with [`Batteries.Marshal.output`](https://ocaml-batteries-team.github.io/batteries-included/hdoc2/Batteries.Marshal.html).

Reply:
It should still be possible to temporary convert anything still present in the weak hashtable into a normal one for marshaling purposes. If everything gets added back in the correct order during unmarshaling, it should be equivalent.

That would still involve some extra work on the hashtable, but it might not require calling `relift` on everything in the solution and needing `relift` to recurse into inner values through multiple layers. And if it works, then it would also allow removing `relift` from the interface, which is good for cleanup.